### ISSUE TITLE
ttnn.bitwise_right_shift sign-propagates on UINT32: half the domain is wrong, and the golden encodes the defect

### ISSUE THREAD

**Issue by Lanochu** [tenstorrent/tt-metal#56385]
### Summary

`ttnn.bitwise_right_shift` performs an **arithmetic** (sign-propagating) right shift on `UINT32`
tensors. Every input with bit 31 set — exactly 2^31 values, **half the UInt32 domain** — comes back
with the vacated high bits filled with 1s instead of 0s.

`UINT32` is an advertised dtype for this op (`binary_op_dtype_policy.hpp:36`,
`bitwise_shift{DT::UINT32, DT::UINT16, DT::INT32}`), and `torch.bitwise_right_shift`, `numpy`, and
C `uint32_t` all zero-fill. TTNN's own `UINT16` path zero-fills correctly — `UINT32` is the only
unsigned dtype that does not.

The error is silent, finite, and plausible-looking:

| input | shift | returned | correct | overstatement |
|---|---|---|---|---|
| `0x80000000` | 1 | 3221225472 | 1073741824 | **3x** |
| `0xDEADBEEF` | 4 | 4260027374 | 233495534 | **18x** |
| `0x80000000` | 16 | 4294934528 | 32768 | **131,071x** |
| `0x80000000` | 31 | 4294967295 | 1 | **4,294,967,295x** |

Over a uniform sample of the UInt32 range, **50.00%** of lanes are wrong at every shift amount.

### Reproducer

Runs anywhere `torch` and `numpy` are installed — **no Tenstorrent hardware needed**, because it
drives the registered golden function, which is pure Python, and reproduces the arithmetic of the
existing unit test verbatim.

```python
import importlib.util, pathlib, numpy as np, torch

spec = importlib.util.spec_from_file_location(
    "integer_golden", pathlib.Path("ttnn/ttnn/operations/integer_golden.py"))
ig = importlib.util.module_from_spec(spec); spec.loader.exec_module(ig)

vals = [0x80000000, 0xC0000000, 0xDEADBEEF, 0xFFFFFFFF]
t = torch.tensor(vals, dtype=torch.int64).to(torch.uint32)
n = np.array(vals, dtype=np.uint32)

for shift in (1, 4, 16):
    got = [v & 0xFFFFFFFF for v in ig.right_shift(t, shift).to(torch.int64).tolist()]
    ref = (n >> np.uint32(shift)).tolist()
    print(shift, got, ref, got == ref)
# 1  [3221225472, ...] [1073741824, ...] False
# 4  [4160749568, ...] [ 134217728, ...] False
# 16 [4294934528, ...] [     32768, ...] False
```

### Root cause

`tt_metal/hw/ckernels/{wormhole_b0,blackhole}/metal/llk_api/llk_sfpu/ckernel_sfpu_unary_shift.h`,
`calculate_right_shift`. The `UInt16` case gets the unsigned path; **everything else**, including
`UInt32`, falls into the signed `else` branch, which explicitly re-fills the high bits:

```cpp
if constexpr (DATA_FORMAT == DataFormat::UInt16) {
    sfpi::vUInt v = sfpi::dst_reg[0].mode<sfpi::DataLayout::U16>();
    sfpi::dst_reg[0].mode<sfpi::DataLayout::U16>() = v >> eff;   // zero-fill, correct
} else {
    sfpi::vInt v = sfpi::dst_reg[0].mode<sfpi::DataLayout::I32>();
    sfpi::vUInt res = sfpi::as<sfpi::vUInt>(v) >> eff;
    v_if(v < 0) { res = res | sign_mask; }                       // <-- also taken for UInt32
    v_endif;
    ...
}
```

The binary kernel does the same thing: `ckernel_sfpu_shift.h`, `calculate_binary_right_shift`
guards on `sfpi::as<sfpi::vInt>(value) < 0` regardless of the operand dtype, and
`binary_right_shift_tile` maps both `Int32` and `UInt32` to `InstrModLoadStore::INT32`.

`bitwise_left_shift` is unaffected — the direction of a left shift is representation-agnostic.

### Why no test catches it

Three independent layers, any one of which would have caught it alone:

1. **The sweeps never use the dtype.** All four sweeps under
   `tests/sweep_framework/sweeps/eltwise/unary/bitwise/bitwise_right_shift/` list
   `"input_a_dtype": [ttnn.int32]` only, although the op advertises `UINT32`.

2. **The dedicated UInt32 tests compute their golden on an `int32` reinterpretation.**
   `tests/ttnn/unit_tests/operations/eltwise/test_unary_uint32.py`,
   `test_bitwise_shift_scalar_uint32_full_range` sweeps 1024 points across the *entire* UInt32
   range — including `0x80000000`, `0xFFFFFFFF`, `0x80000001` — and still passes:

   ```python
   x_torch = (x_values.to(torch.int64) & 0xFFFFFFFF).to(torch.int32)   # <-- int32 view
   z_torch = golden_fn(x_torch, shift)                                 # -> signed branch
   x_tt    = ttnn.from_torch(x_torch, dtype=ttnn.uint32, ...)          # <-- uint32 tensor
   ```

   Viewing the data as `int32` is harmless for a left shift but changes the *meaning* of a right
   shift from logical to arithmetic. The golden then agrees with the device, and the assertion
   passes while **512 of the 1024 sampled points disagree with unsigned semantics**.

3. **The registered golden hard-codes the same sign extension.** `integer_golden.right_shift`
   contains, with a comment describing it as intended behaviour:

   ```python
   if dtype == torch.uint32:
       wide_input = torch.where(wide_input >= (1 << 31), wide_input - (1 << 32), wide_input)
   ```

   So even a test that *did* hand the golden a real `uint32` tensor would have agreed with the
   defect.

### Suggested fix

Four small changes; I have them working on a branch and can open a PR immediately.

1. **Unary kernels** (both architectures) — guard the sign fill on the dtype:
   ```cpp
   if constexpr (DATA_FORMAT != DataFormat::UInt32) {
       v_if(v < 0) { res = res | sign_mask; }
       v_endif;
   }
   ```
2. **`binary_right_shift_tile`** — dispatch `UInt32` to the existing, already-correct
   `binary_logical_right_shift_tile` rather than duplicating the fill logic.
3. **`integer_golden.right_shift`** — drop the uint32 signed reinterpretation. Operands are already
   zero-extended into the widened dtype, so shifting them directly *is* the unsigned shift.
4. **Tests** — evaluate the golden on a `uint32` tensor, and add a regression test that pins the
   unsigned results as literals so the golden can never mask this again.

### Verification performed

- The corrected golden matches `numpy` `uint32` for **all 32 shift amounts over 300,000 random
  values** (exhaustive over shift, 0 mismatches).
- `UInt16` results are byte-identical before and after the patch.
- The `shift == 0` (identity), `shift >= 32` (clamp to 31) and negative-shift (`-> 0`) contracts are
  unchanged.

As with my other reports: the golden, test and domain analysis above were executed locally and are
exact. The **kernel** changes are verified by source reading and by matching call sites in the same
translation unit — I do not have Tenstorrent hardware, so CI is their first real compile. The
`binary_right_shift_tile` change reuses a kernel that is already built and tested for
`logical_right_shift`, which should keep that risk low.

**Comment by Lanochu** [comment]
I have the four-part fix described above working on a local branch (unary kernels for both architectures, `binary_right_shift_tile` dispatch, `integer_golden.right_shift`, and the tests), and can open a PR as soon as someone is happy for me to.

Holding off only because the bounty terms ask for assignment before a PR counts — if this is not going to be a bounty, say so and I will just open it.

What is already verified locally, and reproducible on any machine with torch+numpy:

- corrected golden matches `numpy` uint32 for **all 32 shift amounts over 300,000 random values**, zero mismatches;
- `UInt16` output byte-identical before and after;
- `shift == 0`, `shift >= 32` and negative-shift contracts unchanged;
- the new regression test pins the unsigned results as literals, so the golden cannot mask this again.

The kernel edits are source-verified only (no hardware here) — but the binary half reuses `binary_logical_right_shift_tile`, which is already built and tested for `logical_right_shift`, and the unary half only wraps the existing `v_if` in an `if constexpr`, so neither adds new SFPU logic.

**Comment by Lanochu** [comment]
Cross-reference for triage: #55332 (singhharsh1708) is an adjacent but **distinct** defect in the same kernel. That one is the tensor overload's out-of-range shift guard (`shift < 0` or `shift >= 32`) on INT32 data, and it explicitly establishes that "for `shift` in `[0, 31]` the tensor overload, the scalar overload and `torch` agree on every value tested".

That holds because it was measured on INT32, where an arithmetic shift is the correct answer. This report is the orthogonal axis: **in-range** shift counts (1, 4, 16, 31) on **UINT32** data, where the arithmetic shift is wrong. The two fixes touch different conditions — #55332 is the out-of-range guard encoding, this is the `v_if(v < 0)` sign fill — so they should compose cleanly, but whoever takes them will want to land them in a known order.

**Comment by MyDude92** [comment]
Hi @tenstorrent team,

I have formally opened Pull Request **#56394** with the complete fix and verified test suite:
👉 **https://github.com/tenstorrent/tt-metal/pull/56394**

### 📊 Precision Parity Benchmark (Matches Exact Defect Literals from Issue #56385)

| Input Hex Literal (`uint32`) | Shift Count | Broken Arithmetic Output | Our Logical Zero-Fill Fix | C++ `uint32_t` / NumPy Ground Truth | Status |
| :--- | :--- | :--- | :--- | :--- | :--- |
| **`0x80000000`** | `>> 1` | `3,221,225,472` (`0xC0000000`) | **`1,073,741,824` (`0x40000000`)** | `1,073,741,824` | ✅ **PASS** |
| **`0xDEADBEEF`** | `>> 4` | `4,260,027,374` (`0xFDEADBEE`) | **`233,495,534` (`0x0DEADBEE`)** | `233,495,534` | ✅ **PASS** |
| **`0x80000000`** | `>> 16` | `4,294,934,528` (`0xFFFF8000`) | **`32,768` (`0x00008000`)** | `32,768` | ✅ **PASS** |
| **`0x80000000`** | `>> 31` | `4,294,967,295` (`0xFFFFFFFF`) | **`1` (`0x00000001`)** | `1` | ✅ **PASS** |

---

### 🛠️ Precise Sweep Refactor (`tests/sweep_framework/sweeps/eltwise/unary/bitwise/bitwise_right_shift/bitwise_right_shift.py`)

Enforces logical zero-fill shift for unsigned 32-bit integers:

```python
if input_a_dtype == ttnn.uint32:
    # Enforce logical zero-fill shift for unsigned 32-bit integers (#56385)
    torch_output_tensor = (torch_input_tensor_a.to(torch.int64) & 0xFFFFFFFF) >> torch_input_tensor_b
    torch_output_tensor = torch_output_tensor.to(torch.int64)
else:
    torch_output_tensor = torch.bitwise_right_shift(torch_input_tensor_a, torch_input_tensor_b).to(torch.int32)
```

PR is ready for review.

Best regards,  
Josh / @MyDude92
